net/src/connection/net_connection.rs

```diff
@@ -1,7 +1,40 @@
 use super::{protocol::Protocol, NetResult};
+use parking_lot::RwLock;
+use std::{fmt, sync::Arc};
 
 /// closure for processing a Protocol message received from the network
-pub type NetHandler = Box<FnMut(NetResult<Protocol>) -> NetResult<()> + Send>;
+#[derive(Clone, Serialize)]
+pub struct NetHandler {
+    #[serde(skip)]
+    closure: Arc<RwLock<Box<FnMut(NetResult<Protocol>) -> NetResult<()> + Send + Sync>>>,
+}
+
+impl NetHandler {
+    pub fn new(c: Box<FnMut(NetResult<Protocol>) -> NetResult<()> + Send + Sync>) -> NetHandler {
+        NetHandler {
+            closure: Arc::new(RwLock::new(c)),
+        }
+    }
+
+    pub fn handle(&mut self, message: NetResult<Protocol>) -> NetResult<()> {
+        (Arc::get_mut(&mut self.closure).ok_or(failure::err_msg(
+        let mut lock = self.closure.write();
+        ))?)(message)
+        (&mut *lock)(message)
+    }
+}
+
+impl PartialEq for NetHandler {
+    fn eq(&self, _: &NetHandler) -> bool {
+        false
+    }
+}
```

Comment:
These trait impls are needed since `NetHandler` is now passed as parameter to the init action.
Yes, that is the same reason for making `NetHandler` `Send` (which needs `crossbeam` for the channels in the tests).

Actions are passed through a channel into the redux main loop -> hence `Send`.
Actions are also stored in `HashMap`s in the state -> hence `PartialEq`.

Hm, thanks for asking about the handlers never being equal. I guess that is the reason I still see blocking actions. Some are waiting for a specific action to appear in the state history (which I consider also tech-debt). With this false those futures will never resolve. I'm looking into this. Guess we can't have `eq` always be false...

Reply:
@lucksus once you say this code related to `PartialEq` trait is ready, I'll approve.